src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -2385,5 +2389,223 @@ .locals init (bool V_0)
                 }
             }
         }
+
+        [Fact, WorkItem(35584, "https://github.com/dotnet/roslyn/issues/35584")]
+        public void MatchToTypeParameterUnbox_01()
+        {
+            var source = @"
+class Program
+{
+    public static void Main() => System.Console.WriteLine(P<int>(0));
+    public static string P<T>(T t) => (t is object o) ? o.ToString() : string.Empty;
+}
+";
+            var expectedOutput = @"0";
+            foreach (var options in new[] { TestOptions.DebugExe, TestOptions.ReleaseExe })
+            {
+                var compilation = CreateCompilation(source, options: options);
+                compilation.VerifyDiagnostics();
+                var compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+                if (options.OptimizationLevel == OptimizationLevel.Debug)
+                {
+                    compVerifier.VerifyIL("Program.P<T>",
+@"{
+  // Code size       24 (0x18)
+  .maxstack  1
+  .locals init (object V_0) //o
+  IL_0000:  ldarg.0
+  IL_0001:  box        ""T""
+  IL_0006:  stloc.0
+  IL_0007:  ldloc.0
+  IL_0008:  brtrue.s   IL_0011
+  IL_000a:  ldsfld     ""string string.Empty""
+  IL_000f:  br.s       IL_0017
+  IL_0011:  ldloc.0
+  IL_0012:  callvirt   ""string object.ToString()""
+  IL_0017:  ret
+}
+");
+                }
+                else
+                {
+                    compVerifier.VerifyIL("Program.P<T>",
+@"{
+  // Code size       23 (0x17)
+  .maxstack  1
+  .locals init (object V_0) //o
+  IL_0000:  ldarg.0
+  IL_0001:  box        ""T""
+  IL_0006:  stloc.0
+  IL_0007:  ldloc.0
+  IL_0008:  brtrue.s   IL_0010
+  IL_000a:  ldsfld     ""string string.Empty""
+  IL_000f:  ret
+  IL_0010:  ldloc.0
+  IL_0011:  callvirt   ""string object.ToString()""
+  IL_0016:  ret
+}
+");
+                }
+            }
+        }
+
+        [Fact, WorkItem(35584, "https://github.com/dotnet/roslyn/issues/35584")]
+        public void MatchToTypeParameterUnbox_02()
+        {
+            var source =
+@"using System;
+ class Program
+{
+    public static void Main()
+    {
+        var nonGenericfoo = new NonGenericFoo(0);  // no problem
+        var genericFoo = new GenericFoo<int>(0); // Kaboom exception thrown
```

Comment:
Why would this throw? I also don't see it throwing in the output

Reply:
It doesn't throw because the bug is fixed.  Should I revise the comments?

Reply:
Makes sense - seems like comment rot otherwise. They were only added to call out where the bug occurred, so now that it no longer occurs, they're outdated, and their presence in the test code suggests that an exception is actually expected. I'd suggest either removing them, or adjusting them to indicate that a bug *used* to occur there.

Reply:
Yup, this reads very strangely right now